### ISSUE TITLE
question: Best approach to update all objects on cluster-wide event (new namespace)

### ISSUE THREAD

**Issue by zakkg3** [zalando-incubator/kopf#365]
## Question

HI! :)

I am working on a cross-namespaces resource that i need to create and maintain in every namespace that match some criteria.
I am stuck on reacting to new namespaces and run the kopf.on.resume or .on.update handler for all existing objects in all namespaces (crds)  to re-sync all the objects with the new namespace.
something like this:

```python
def _new_match_ns(**kwargs):
    return kwargs['event']['type'] == 'ADDED'

@kopf.on.event('', 'v1', 'namespaces',when=_new_match_ns)
async def namespace_watcher(patch,logger,meta,body,event,**kwargs):
    logger.info(f"New namespace created: {meta['name']} re-syncing")
       # for obj in all_objects: a_handler_fn(obj) #re-sync 
```
but this will run N ( number of objects) times?, not very optimal.

maybe `asyncio.create_task(ns_watcher(**kwargs)` on.create and/or on.resume, so i have a async namespace watcher per object . On new namespace -> re sync .this object. But is this right? make sense to write my own watcher in this fn?, Didnt manage to make it work like this and i suspect kopf provides a better way to do that...

I been reading the docs in the last days, kopf.execute / idempotence, sub handlers, and so on . but I don't see whats the best approach here. 

thanks!!!

## Checklist

- [x] I have read the [documentation](https://kopf.readthedocs.io/en/latest/) and searched there for the problem
- [x] I have searched in the [GitHub Issues](https://github.com/zalando-incubator/kopf/issues?utf8=%E2%9C%93&q=) for similar questions

## Keywords

/ edit 
i found how to kill the threads here https://github.com/zalando-incubator/kopf/issues/122
but still dont know if this is the best way... and because my thread is watching for ns events when deleting it get stuck there, it never goes to the while loop 'while shouldrun:' :/

**Comment by nolar** [comment]
You can try using on.create/on.update/on.delete handlers for namespaces. For that, you probably will need `kopf>=0.27rc6` (or 0.27 once released; see [release notes](https://github.com/zalando-incubator/kopf/releases)), as it provides ways of configuring operator's identity — to not collide with other Kopf-based operators serving the same object.

Specifically, Kopf will put a finalizer on the namespaces, so they will no be actually deleted until your operator executes all deletion handlers.

Similarly, multiple iterations will be possible both for creation & deletion of the namespaces, if there are multiple handlers or sub-handlers or errors or other reasons to make multiple cycles.

This seems like the most straightforward way.

Using on.event for this kind of task is not reliable: no error retries, no sub-handlers. It is the lowest-level handler possible — basically, what Kubernetes sends, is passed to on.event "as is" (almost), fire-and-forget style. Specifically, `ADDED` is sent as such only when the resource appears in Kubernetes once during watching; all other times it is `MODIFIED`, even if the multi-step creation handling is not finished by Kopf; or `None` if the resource is noticed during initial listing on startup. Similarly, event type `DELETED` is sent not when the resource is marked for deletion, but when it is fully gone, and it is too late to react.

---

I didn't get the idea with the watchers, but if you look into creating background tasks, take a look at [daemons](https://kopf.readthedocs.io/en/latest/daemons/) & [timers](https://kopf.readthedocs.io/en/latest/timers/) (also available since 0.27 release candidates). 

Daemons run as long as the resource exists: both in threads (sync functions) and asyncio tasks (async functions), and have a helper for graceful termination ([`stopped`](https://kopf.readthedocs.io/en/latest/kwargs/#std-kwarg-stopped)). 

Timers are a special kind of daemons, which sleep most of the time and wake up regularly. They are basically a replacement for daemons with `while not stopped: doit(); sleep(N)` logic.

---

As a rule of thumb, writing your own watchers can be not such a good idea: the logic of watching the resources is complicated: dealing with TCP streaming connections, timeouts, authentication, interpreting ADDED/MODIFIED/DELETED event types, ERROR event types, etc.

Kopf itself is basically a watcher that marshals events to the functions, and deals with all that complexity of underlying Kubernetes API. So, it would be wise to leverage that, and only implement those handler functions to keep a global state of known namespaces/resources.

---

The logic of cloning the resources to new namespaces is not directly supported by Kopf. It requires joining the existing namespaces with existing resources, all-to-all, many-to-many. 

The operators are by concept event-driven, or, better say, state-driven. But the handlers are invoked for one individual resource only. So, to implement all-to-all mapping, you have to keep a full list of at each of these resources (your own + namespaces) in memory, and use them as a kind of "template" when something happens on another resource:

I.e., when a namespace is created, iterate over all known "template" objects of your kind, and clone them to that specific namespace. When a new "template" object is created, iterate over all known/matching namespaces, and make a clone of this specific object to each of these namespaces.

Storing such state in operator's global variables sounds good enough.

Using on.create/on.delete handlers will ensure that the steps are not "lost" due to errors, and all handlers/sub-handlers are indeed executed. This goes at the cost of storing some temporary information in the namespaces' annotations.

---

These are just a few "blind answers" to an extent of my understanding of the question. If you provide more information on your proposed solution, I can be more specific with the suggestions.

**Comment by zakkg3** [comment]
@nolar Thanks Its just what i needed!
